pkg/cloudprovider/providers/openstack/openstack.go

```diff
@@ -225,37 +230,98 @@ func configFromEnv() (cfg Config, ok bool) {
 		cfg.Global.DomainName = os.Getenv("OS_USER_DOMAIN_NAME")
 	}
 
-	ok = cfg.Global.AuthURL != "" &&
-		cfg.Global.Username != "" &&
-		cfg.Global.Password != "" &&
-		(cfg.Global.TenantID != "" || cfg.Global.TenantName != "" ||
-			cfg.Global.DomainID != "" || cfg.Global.DomainName != "" ||
-			cfg.Global.Region != "" || cfg.Global.UserID != "" ||
-			cfg.Global.TrustID != "")
-
 	cfg.Metadata.SearchOrder = fmt.Sprintf("%s,%s", configDriveID, metadataID)
 	cfg.Metadata.DHCPDomain = "novalocal"
 	cfg.BlockStorage.BSVersion = "auto"
-
-	return
 }
 
-func readConfig(config io.Reader) (Config, error) {
-	if config == nil {
-		return Config{}, fmt.Errorf("no OpenStack cloud provider config file given")
+func configFromSecret(cfg *Config, kubeClient clientset.Interface, name string, namespace string) error {
+	if name == "" {
+		name = "openstack"
+	}
+	if namespace == "" {
+		namespace = "kube-system"
+	}
+	secret, err := kubeClient.CoreV1().Secrets(namespace).Get(name, metav1.GetOptions{})
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+	} else {
+		for name, data := range secret.Data {
+			value := string(data)
+			switch name {
+			case "auth-url":
+				cfg.Global.AuthURL = value
+			case "username":
+				cfg.Global.Username = value
+			case "user-id":
+				cfg.Global.UserID = value
+			case "password":
+				cfg.Global.Password = value
+			case "tenant-id":
+				cfg.Global.TenantID = value
+			case "tenant-name":
+				cfg.Global.TenantName = value
+			case "trust-id":
+				cfg.Global.TrustID = value
+			case "domain-id":
+				cfg.Global.DomainID = value
+			case "domain-name":
+				cfg.Global.DomainName = value
+			case "region":
+				cfg.Global.Region = value
+			case "ca-file":
+				cfg.Global.CAFile = value
```

Comment:
does it make sense to reference a filepath on the apiserver machine via a config value loaded from an API resource like a secret?

Reply:
Ack removing ca-file and auth-url